File: Crio/crio_programming_interview_problems-master/EditDistance/EditDistance.cpp
#include <bits/stdc++.h> 
using namespace std;  
// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
int min(int x, int y, int z) 
{ 
	return min(min(x, y), z); 
} 
// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
int editDistance(string str1, string str2) 
{ 
// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
	int m = str1.length();
	int n = str2.length();
	int dp[m + 1][n + 1]; 
	for (int i = 0; i <= m; i++) { 
		for (int j = 0; j <= n; j++) { 
			if (i == 0) 
				dp[i][j] = j;
			else if (j == 0) 
				dp[i][j] = i;
			else if (str1[i - 1] == str2[j - 1]) 
				dp[i][j] = dp[i - 1][j - 1]; 
			else
				dp[i][j] = 1 + min(dp[i][j - 1],dp[i - 1][j],dp[i - 1][j - 1]);
		} 
	} 
	return dp[m][n]; 
// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
} 
int main() 
{
	string str1,str2;
	cin>>str1;
	cin>>str2;
	cout << editDistance(str1, str2);
}

File: Crio/crio_programming_interview_problems-master/EditDistance/EditDistance.py
import sys
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
dp=list()
cnt = 0
def cal(word1,word2,n,m):
	if n==0:
		return m
	if m==0:
		return n
	global dp
	global cnt
	if dp[n-1][m-1] != -1:
		return dp[n-1][m-1]
	cnt+=1
	if word1[n-1]== word2[m-1]:
		dp[n-1][m-1] = cal(word1,word2,n-1,m-1)
		return dp[n-1][m-1]  
	x= cal(word1,word2,n,m-1)
	y= cal(word1,word2,n-1,m)
	z= cal(word1,word2,n-1,m-1)

	dp[n-1][m-1]=1+min(x,y,z)
	return dp[n-1][m-1]
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

def editDistance(word1,word2):
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
	n=len(word1)
	m=len(word2)
	global dp
	global cnt
	dp =[[-1 for x in range(m+1)] for y in range(n+1)] 
	ans=cal(word1,word2,n,m)
	return ans
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

def main():
	word1=input()
	word2=input()
	sys.setrecursionlimit(1000000)
	ans=editDistance(word1,word2)
	print(ans)


if __name__ == '__main__':
	main()
